blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,394 @@
+- Built-in Spans: Leveraging Next.js’s automatic spans and debugging with NEXT_OTEL_VERBOSE
+- Exception Tracking: Capturing errors within traces
+- Simplifying Setup with @vercel/otel
+
+By the end of this tutorial, you’ll have OpenTelemetry integrated into your Next.js project with traces, logs, and metrics exported to your preferred backend.
+
+### Installing OpenTelemetry in Next.js
+
+To get started, install the necessary OpenTelemetry dependencies:
+
+```bash
+yarn add @opentelemetry/api @opentelemetry/sdk-node @opentelemetry/instrumentation-http @opentelemetry/instrumentation-fetch @opentelemetry/exporter-trace-otlp-grpc @opentelemetry/exporter-metrics-otlp-grpc @opentelemetry/resources @opentelemetry/semantic-conventions
+```
+
+This setup includes the core OpenTelemetry API, SDK, HTTP and Fetch instrumentations, and OTLP exporters for traces and metrics.
+
+### Setting Up OpenTelemetry SDK
+
+Create a new file `otel.js` at the root of your Next.js project:
+
+```javascript
+import { NodeSDK } from '@opentelemetry/sdk-node';
+import { ConsoleSpanExporter } from '@opentelemetry/sdk-trace-base';
+import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
+import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-grpc';
+import { Resource } from '@opentelemetry/resources';
+import { SEMRESOURCENAME } from '@opentelemetry/semantic-conventions';
+import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
+import { FetchInstrumentation } from '@opentelemetry/instrumentation-fetch';
+
+const sdk = new NodeSDK({
+  resource: new Resource({
+    [SEMRESOURCENAME.SERVICE_NAME]: 'nextjs-app',
+  }),
+  traceExporter: new OTLPTraceExporter(), 
+  metricExporter: new OTLPMetricExporter(), 
+  instrumentations: [new HttpInstrumentation(), new FetchInstrumentation()],
+});
+
+sdk.start();
+console.log('OpenTelemetry initialized');
+```
+
+To trigger this file to run when the app starts, you can invoke it from the Next.js magic `instrumentation.js` file.
+
+```javascript
+import './otel';
+```
+
+The [instrumentation.js file](https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation) is automatically detected by Next.js and will run when the app starts. Before Next.js 15, the instrumentation is experimental, so you will have to enable it explicitly:
+
+```javascript
+module.exports = {
+  experimental: {
+    instrumentationHook: true,
+  },
+};
+```
+
+
+### Enabling OpenTelemetry in Next.js
+
+Automatic Spans in Next.js
+
+Next.js has built-in OpenTelemetry spans for various parts of the application, including:
+- API routes (pages/api or app/api)
+- Page router (Pages Directory)
+- App router (App Directory)
+
+To enable verbose OpenTelemetry logging, set the following environment variable:
+
+```bash
+NEXT_OTEL_VERBOSE=1
+```
+
+With this enabled, Next.js will provide detailed span information in the logs, making it easier to debug and understand request flow.
```

Comment:
What do you think about sharing an example of why this is useful?

Reply:
added with a comparison